### PR TITLE
[Feature] Support ISA module

### PR DESCRIPTION
**Pull request by LayneH**

### CODE REVIEW

File: mmseg/models/decode_heads/__init__.py [hunk 1]
```diff
@@ -13,7 +14,18 @@
...
+    'NLHead',
+    'GCHead',
+    'CCHead',
+    'UPerHead',
+    'DepthwiseSeparableASPPHead',
+    'ANNHead',
+    'DAHead',
+    'OCRHead',
+    'EncHead',
+    'ISAHead',
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/__init__.py:30]: The comma could be removed.

File: mmseg/models/decode_heads/isa_head.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+@HEADS.register_module()
+class ISAHead(BaseDecodeHead):
+    """Interlaced Sparse Self-Attention for Semantic Segmentation.
+
+    This head is the implementation of `ISA
+    <https://arxiv.org/abs/1907.12273>`.
+
+    Args:
+        isa_channels (int): The channels of ISA Module.
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/isa_head.py:62]: Missing arguments.

File: mmseg/models/decode_heads/isa_head.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            act_cfg=self.act_cfg)
+
+    def forward(self, inputs):
+        """Forward function."""
+        x_ = self._transform_inputs(inputs)
+        x = self.in_conv(x_)
+
+        n, c, h, w = x.size()
+        P_h, P_w = self.down_factor
+        Q_h, Q_w = math.ceil(h / P_h), math.ceil(w / P_w)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/isa_head.py:104]: We may use more readable variable names.

File: mmseg/models/decode_heads/isa_head.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            padding = (pad_w // 2, pad_w - pad_w // 2, pad_h // 2,
+                       pad_h - pad_h // 2)
+            feat = F.pad(x, padding)
+        else:
+            feat = x
+
+        # global relation
+        feat = feat.view(n, c, Q_h, P_h, Q_w, P_w)
+        feat = feat.permute(0, 3, 5, 1, 2, 4).reshape(-1, c, Q_h, Q_w)
+        feat = self.global_relation(feat)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/isa_head.py:116]: Add comments about shape.

File: mmseg/models/decode_heads/isa_head.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        # local relation
+        feat = feat.view(n, P_h, P_w, c, Q_h, Q_w)
+        feat = feat.permute(0, 4, 5, 3, 1, 2).reshape(-1, c, P_h, P_w)
+        feat = self.local_relation(feat)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/isa_head.py:121]: Add comments about shape.

File: mmseg/models/decode_heads/isa_head.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+        feat = feat.view(n, Q_h, Q_w, c, P_h, P_w)
+        feat = feat.permute(0, 3, 1, 4, 2, 5).reshape(n, c, P_h * Q_h,
+                                                      P_w * Q_w)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/isa_head.py:125]: Add comments about shape.

**Comment by Junjun2016** [comment]
Hi @LayneH 
Could you please update this PR if your time is available?

**Comment by LayneH** [comment]
Hi, @xvjiarui  @Junjun2016,

This PR has been updated according to the comments.

**Comment by Junjun2016** [comment]
Hi @LayneH 
Thanks for your contribution.
Could you please resolve the conflict?
And we will review it again.

File: configs/_base_/models/isanet_r50-d8.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        concat_input=False,
+        dropout_ratio=0.1,
+        num_classes=19,
+        norm_cfg=norm_cfg,
+        align_corners=False,
+        loss_decode=dict(
+            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4)))
+# model training and testing settings
+train_cfg = dict()
+test_cfg = dict(mode='whole')
```
Comment (Junjun2016) [inline configs/_base_/models/isanet_r50-d8.py:45]: `train_cfg` and `test_cfg` have been moved to model.

File: configs/isanet/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
+# Interlaced Sparse Self-Attention for Semantic Segmentation
+
+## Introduction
+```
```
Comment (Junjun2016) [inline configs/isanet/README.md:4]: Add tag `<!-- [ALGORITHM] --> `to README, please refer to the README of OCRNet.

File: mmseg/models/decode_heads/isa_head.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+from ..utils import SelfAttentionBlock as _SelfAttentionBlock
+from .decode_head import BaseDecodeHead
+
+
+class SelfAttentionBlock(_SelfAttentionBlock):
+    """Self-Attention Module.
+
+    Args:
+        in_channels (int): Input channels of key/query feature.
+        channels (int): Output channels of key/query transform.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/isa_head.py:17]: Missing docstring for `conv_cfg`, `norm_cfg`,  and `act_cfg`.

File: mmseg/models/decode_heads/isa_head.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        context = super(SelfAttentionBlock, self).forward(x, x)
+        return self.output_project(context)
+
+
+@HEADS.register_module()
+class ISAHead(BaseDecodeHead):
+    """Interlaced Sparse Self-Attention for Semantic Segmentation.
+
+    This head is the implementation of `ISA
+    <https://arxiv.org/abs/1907.12273>`.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/isa_head.py:59]:
```suggestion
    <https://arxiv.org/abs/1907.12273>`_.
```

Comment (Junjun2016) [reply mmseg/models/decode_heads/isa_head.py:59]: The citation style is not correct.
Please refer to https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/mobilenet_v3.py#L19.

**Review by Junjun2016 — COMMENTED** [review]
Please add unittests for `mmseg/models/decode_heads/isa_head.py`.
We will benchmark it recently.

**Comment by Junjun2016** [comment]
`configs/_base_/models/.isanet_r50-d8.py.swp` should be removed.

**Comment by Junjun2016** [comment]
> Please add unittests for `mmseg/models/decode_heads/isa_head.py`.
> We will benchmark it recently.

Refer to https://github.com/open-mmlab/mmsegmentation/blob/master/tests/test_models/test_heads/test_ocr_head.py.

File: mmseg/models/decode_heads/isa_head.py [hunk 11]
```diff
@@ -15,6 +15,11 @@ class SelfAttentionBlock(_SelfAttentionBlock):
     Args:
         in_channels (int): Input channels of key/query feature.
         channels (int): Output channels of key/query transform.
+        share_key_query (bool): Whether share projection weight between key
+            and query projection.
+        conv_cfg (dict|None): Config of conv layers.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/isa_head.py:20]:
```suggestion
        conv_cfg (dict | None): Config of conv layers.
```

File: mmseg/models/decode_heads/isa_head.py [hunk 12]
```diff
@@ -15,6 +15,11 @@ class SelfAttentionBlock(_SelfAttentionBlock):
...
+        norm_cfg (dict|None): Config of norm layers.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/isa_head.py:21]:
```suggestion
        norm_cfg (dict | None): Config of norm layers.
```

File: mmseg/models/decode_heads/isa_head.py [hunk 13]
```diff
@@ -15,6 +15,11 @@ class SelfAttentionBlock(_SelfAttentionBlock):
...
+        act_cfg (dict|None): Config of activation layers.
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/isa_head.py:22]:
```suggestion
        act_cfg (dict | None): Config of activation layers.
```

**Comment by Junjun2016** [comment]
Hi @LayneH 
The unittests are failed.
Please also refer to https://github.com/open-mmlab/mmsegmentation/blob/master/tests/test_models/test_heads/test_dnl_head.py.

**Comment by LayneH** [comment]
Hi @Junjun2016,
The issue has been fixed.

**Comment by Junjun2016** [comment]
> Hi @Junjun2016,
> The issue has been fixed.

Ok, we will add benchmarks for ISA recently.

File: configs/isanet/isanet_r50-d8_512x512_160k_ade20k.py [hunk 14]
```diff
@@ -0,0 +1,7 @@
+_base_ = [
+    '../_base_/models/isanet_r50-d8.py', '../_base_/datasets/ade20k.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_160k.py'
+]
+model = dict(
+    decode_head=dict(num_classes=150), auxiliary_head=dict(num_classes=150))
+test_cfg = dict(mode='whole')
```
Comment (Junjun2016) [inline configs/isanet/isanet_r50-d8_512x512_160k_ade20k.py:7]: `train_cfg` and `test_cfg` have been moved to `model`, and` test_cfg = dict(mode='whole')` has already defined in `configs/_base_/models/isanet_r50-d8.py`, so this line is redundant.

File: configs/isanet/isanet_r50-d8_512x512_80k_ade20k.py [hunk 15]
```diff
@@ -0,0 +1,7 @@
+_base_ = [
+    '../_base_/models/isanet_r50-d8.py', '../_base_/datasets/ade20k.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_80k.py'
+]
+model = dict(
+    decode_head=dict(num_classes=150), auxiliary_head=dict(num_classes=150))
+test_cfg = dict(mode='whole')
```
Comment (Junjun2016) [inline configs/isanet/isanet_r50-d8_512x512_80k_ade20k.py:7]: `train_cfg` and `test_cfg` have been moved to `model`, and` test_cfg = dict(mode='whole')` has already defined in `configs/_base_/models/isanet_r50-d8.py`, so this line is redundant.

File: configs/isanet/isanet_r50-d8_769x769_40k_cityscapes.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+_base_ = [
+    '../_base_/models/isanet_r50-d8.py',
+    '../_base_/datasets/cityscapes_769x769.py', '../_base_/default_runtime.py',
+    '../_base_/schedules/schedule_40k.py'
+]
+model = dict(
+    decode_head=dict(align_corners=True),
+    auxiliary_head=dict(align_corners=True))
+test_cfg = dict(mode='slide', crop_size=(769, 769), stride=(513, 513))
```
Comment (Junjun2016) [inline configs/isanet/isanet_r50-d8_769x769_40k_cityscapes.py:9]: Same.

File: configs/isanet/isanet_r50-d8_769x769_80k_cityscapes.py [hunk 17]
```diff
@@ -0,0 +1,9 @@
+_base_ = [
+    '../_base_/models/isanet_r50-d8.py',
+    '../_base_/datasets/cityscapes_769x769.py', '../_base_/default_runtime.py',
+    '../_base_/schedules/schedule_80k.py'
+]
+model = dict(
+    decode_head=dict(align_corners=True),
+    auxiliary_head=dict(align_corners=True))
+test_cfg = dict(mode='slide', crop_size=(769, 769), stride=(513, 513))
```
Comment (Junjun2016) [inline configs/isanet/isanet_r50-d8_769x769_80k_cityscapes.py:9]: Same.

File: configs/isanet/isanet_r50-d8_769x769_40k_cityscapes.py (hunk 16 above)
Comment (Junjun2016) [reply configs/isanet/isanet_r50-d8_769x769_40k_cityscapes.py:9]: Move `test_cfg` to `model`.

File: configs/isanet/isanet_r50-d8_769x769_80k_cityscapes.py (hunk 17 above)
Comment (Junjun2016) [reply configs/isanet/isanet_r50-d8_769x769_80k_cityscapes.py:9]: Move `test_cfg` to `model`.

File: configs/isanet/isanet_r50-d8_769x769_40k_cityscapes.py (hunk 16 above)
Comment (Junjun2016) [reply configs/isanet/isanet_r50-d8_769x769_40k_cityscapes.py:9]: If `train_cfg` or `test_cfg` is different from default, we should overwrite it.

**Review by Junjun2016 — CHANGES_REQUESTED** [review]

**Comment by Junjun2016** [comment]
Please also fix the conflict.
We are running benchmark.
We will update the results when we finish it.

**Comment by LayneH** [comment]
Hi, the conflicts have been fixed.

**Comment by xiexinch** [comment]
Hi @Junjun2016 
Results updated.

**Comment by LayneH** [comment]
@xiexinch @Junjun2016 
Thank you for the benchmarking, the results seem to be consistent with those in our paper.

**Review by Junjun2016 — CHANGES_REQUESTED** [review]
Please add supported methods in README (English and Chinese).

**Review by Junjun2016 — APPROVED** [review]

**Review by xvjiarui — APPROVED** [review]

**Comment by OpenMMLab-Assistant003** [comment]
Hi @LayneH！First of all, we want to express our gratitude for your significant PR in the MMSegmentation project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤ @LayneH

**Pull request merged**